src/raftstore/store/peer_storage.rs

```diff
@@ -396,43 +392,58 @@ impl PeerStorage {
 
     pub fn snapshot(&self) -> raft::Result<Snapshot> {
         let mut snap_state = self.snap_state.borrow_mut();
-
-        if let SnapState::Snap(_) = *snap_state {
-            match mem::replace(&mut *snap_state, SnapState::Relax) {
-                SnapState::Snap(s) => {
-                    if self.validate_snap(&s) {
-                        return Ok(s);
-                    }
+        let mut tried_cnt = self.snap_tried_cnt.borrow_mut();
+
+        let (mut tried, mut snap) = (false, None);
+        if let SnapState::Generating(ref recv) = *snap_state {
+            tried = true;
+            match recv.try_recv() {
+                Err(TryRecvError::Disconnected) => {}
+                Err(TryRecvError::Empty) => {
+                    return Err(raft::Error::Store(
+                        raft::StorageError::SnapshotTemporarilyUnavailable));
                 }
-                _ => unreachable!(),
+                Ok(s) => snap = Some(s),
             }
         }
 
-        if SnapState::Relax == *snap_state {
-            info!("{} requesting snapshot...", self.tag);
-            self.snap_tried_cnt.store(0, Ordering::Relaxed);
-            *snap_state = SnapState::Generating;
-        } else if SnapState::Failed == *snap_state {
-            let mut snap_tried_cnt = self.snap_tried_cnt.load(Ordering::Relaxed);
-            if snap_tried_cnt >= MAX_SNAP_TRY_CNT {
-                return Err(raft::Error::Store(box_err!("failed to get snapshot after {} times",
-                                                       snap_tried_cnt)));
+        if tried {
+            *snap_state = SnapState::Relax;
+            if let Some(s) = snap {
+                if self.validate_snap(&s) {
+                    *tried_cnt = 0;
+                    return Ok(s);
+                }
             }
-            snap_tried_cnt += 1;
-            warn!("{} snapshot generating failed, retry {} time",
+            warn!("{} snapshot generating failed at {} tried times",
                   self.tag,
-                  snap_tried_cnt);
-            self.snap_tried_cnt.store(snap_tried_cnt, Ordering::Relaxed);
-            *snap_state = SnapState::Generating;
-        } else {
-            return Err(raft::Error::Store(raft::StorageError::SnapshotTemporarilyUnavailable));
+                  *tried_cnt);
+        }
+
+        if SnapState::Relax != *snap_state {
```

Comment:
Check if the peer is in a good state. Other states like `Applying` is invalid and there should be bugs somewhere.